// node-red-contrib-tekosbot-analytics/nodes/tekos-dashbot.js
module.exports = function(RED) {
    "use strict";
    // var DeepstreamClient = require('deepstream.io-client-js');
    // var querystring = require('querystring');
    // var axios = require('axios');
    // var dashbot = require('dashbot')(process.env.DASHBOT_API_KEY_GENERIC,{}).generic;
    // if(process.env.DASHBOT_API_KEY_GENERIC && process.env.DASHBOT_API_KEY_GENERIC != '')
    // 	var dashbot = require('dashbot')(process.env.DASHBOT_API_KEY_GENERIC,{}).generic;

    var apiToken = ''; 
    var dashbot = '';

    function DashbotToken(config){
    	var node = this;
    	RED.nodes.createNode(node, config);
    	// console.log(config)
    	
    			try {
					apiToken = config.token;
					
    				dashbot = require('dashbot')(apiToken,{}).generic;

				} catch(err) {
					console.log(err)
					//node.error(err);
				}
    }
    RED.nodes.registerType("dashbot-app",DashbotToken);

    function msgOutLog(uid, msg){
    	if(apiToken && apiToken != ''){
	    	console.log(msg)

	    	dashbot.logOutgoing({
	    		text: msg.payload.content || msg.payload.text || msg.payload.text[0],
	    		userId: msg.id,
	    		conversationId: msg.uid,
	    		platformJson: msg
	    	})
	    	console.log("msgOutLog-------------------------")
    	}
    }
    // msg.content
    function msgInLog(uid, msg){
    	
    	if(apiToken && apiToken != '')
    	{
	    	dashbot.logIncoming({
	    		text: msg.payload.content || msg.payload.text || msg.payload.text[0],
	    		userId: msg.id,
	    		conversationId: msg.uid,
	    		platformJson: msg
	    	})
    		// console.log("msgInLog-------------------------")
	    	// console.log(msg)

    	}
    }

    function botPause(uid){
    	
    	if(apiToken && apiToken != '')
    	{
	    	dashbot.logIncoming({
	    		text: msg.payload.content || msg.payload.text || msg.payload.text[0],
	    		userId: msg.id,
	    		conversationId: msg.uid,
	    		platformJson: msg
	    	})
    		// console.log("msgInLog-------------------------")
	    	// console.log(msg)

    	}
    }


	/****************************** Dashbot In *******************************/
		
    function DashbotIn(config) {
    	//console.log(config);
        var node = this;
		RED.nodes.createNode(node, config);
				
		//msgOutLog(data.uid,data);
						

        node.on("input", function(msg) {
					// console.log("****************************** Dashbot IN *******************************")
					console.log(msg)
					msgOutLog(msg.uid, msg)
				
						node.send(msg);		        
        });
    }
	RED.nodes.registerType("Dashbot In",DashbotIn);
	
	
	/****************************** Dashbot Out *******************************/
    function DashbotOut(config) {
    	//console.log(config);

        var node = this;
		RED.nodes.createNode(node, config);		
		node.on("input", function(msg) {
		// console.log("****************************** Dashbot Out *******************************")
		// 			console.log(msg)
					msgInLog(msg.uid, msg)
				
						node.send(msg);		        
        });		

    }
	RED.nodes.registerType("Dashbot Out",DashbotOut);
	
	
	/****************************** Register *******************************/		

};
